Add authorized API calls to get list of repo tags

corest opened the pull request:
Towards https://github.com/giantswarm/giantswarm/issues/12052

Code starts to be super messy. It worth to have some generic interface and implement those listing/authorization methods for interface. 

But as we have two repos for now, this covers our use-case. 
Comparing to pulling every tag, this version obviously works much faster and is safe from rate-limiting perspective as we do only one request to get tags

kopiczko commented on pkg/registry/tags.go:
```diff
@@ -0,0 +1,87 @@
+
+	"github.com/giantswarm/microerror"
+)
+
```
I feel like we should have two separate implementations for something like `TagLister` interface and wire a correct one instead of switching on `r.kind` all the time.

corest replied:
I fully agree with you here. I'll try to come up with a valid implementation

corest commented:
I've refactored code a bit to have registry client as interface and moved then all the logic into interface implementation per registry. 
There is still this weird part of credentials duplication, but I like now this better. Even if it still looks terrible :D

kopiczko commented on pkg/azurecr/azurecr.go:
```diff
@@ -0,0 +1,84 @@
+
+func New(c Config) (*AzureCR, error) {
+	httpClient := &http.Client{}
+
```
Config validation is missing

kopiczko commented on pkg/azurecr/azurecr.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+func (d *AzureCR) ListRepositories() ([]string, error) {
+	return []string{}, nil
```
Can we panic here instead? Calling this is a bug I think.

kopiczko commented on pkg/dockerhub/dockerhub.go:
```diff
@@ -0,0 +1,120 @@
+
+func New(c Config) (*DockerHub, error) {
+	httpClient := &http.Client{}
+
```
Config validation.

kopiczko commented on pkg/dockerhub/dockerhub.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+func (d *DockerHub) ListRepositories() ([]string, error) {
+	return []string{}, nil
```
panic

kopiczko commented on pkg/quay/quay.go:
```diff
@@ -0,0 +1,166 @@
+
+	tagLengthLimit = 20
+)
+
+type Quay struct {
+	namespace    string
+	lastModified time.Duration
+
+	httpClient *http.Client
+}
+
+type Config struct {
+	Namespace    string
+	LastModified time.Duration
+}
+
+type Repository struct {
+	Name         string `json:"name"`
+	LastModified int    `json:"last_modified"`
+}
+
+type RepositoriesJSON struct {
+	Repositories []Repository `json:"repositories"`
+}
+
```
I'd prefer to move `Repository` & `RepositoriesJSON` to `types.go`.

```suggestion

type Repository struct {
	Name         string `json:"name"`
	LastModified int    `json:"last_modified"`
}

type RepositoriesJSON struct {
	Repositories []Repository `json:"repositories"`
}

type Config struct {
	Namespace    string
	LastModified time.Duration
}

type Quay struct {
	namespace    string
	lastModified time.Duration

	httpClient *http.Client
}

```

kopiczko commented on pkg/quay/quay.go:
```diff
@@ -0,0 +1,166 @@
+	}
+
+	for _, repo := range data.Repositories {
+		lastModifiedTimestamp := int(time.Now().Add(-1 * q.lastModified).Unix())
+		if repo.LastModified > lastModifiedTimestamp {
```
```suggestion
		lastModifiedTimestamp := time.Now().Add(-1 * q.lastModified).Unix()
		if int64(repo.LastModified) > lastModifiedTimestamp {
```

kopiczko commented on pkg/quay/quay.go:
```diff
@@ -0,0 +1,166 @@
+	{
+		nextEndpoint := endpoint
+		for {
+			resp, err := http.Get(nextEndpoint) // nolint
```
What did linter want?

kopiczko commented on pkg/quay/quay.go:
```diff
@@ -0,0 +1,166 @@
+		for {
+			resp, err := http.Get(nextEndpoint) // nolint
+			if err != nil {
+				return []string{}, microerror.Mask(err)
```
```suggestion
				return nil, microerror.Mask(err)
```

kopiczko commented on pkg/quay/quay.go:
```diff
@@ -0,0 +1,166 @@
+			defer resp.Body.Close()
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return []string{}, microerror.Mask(err)
```
```suggestion
				return nil, microerror.Mask(err)
```

kopiczko commented on pkg/quay/quay.go:
```diff
@@ -0,0 +1,166 @@
+
+			err = json.Unmarshal(body, &tagsData)
+			if err != nil {
+				return []string{}, microerror.Mask(err)
```
```suggestion
				return nil, microerror.Mask(err)
```

kopiczko commented on pkg/quay/quay.go:
```diff
@@ -0,0 +1,166 @@
+	var tags []string
+	{
+		nextEndpoint := endpoint
+		for {
```
```suggestion
		for {
			if nextEndpoint == "" {
				break
			}

```

kopiczko commented on pkg/quay/quay.go:
```diff
@@ -0,0 +1,166 @@
+				}
+			}
+
+			linkHeader := resp.Header.Get("Link")
+			if linkHeader != "" {
+				nextEndpoint = fmt.Sprintf("%s%s", registryEndpoint, getLink(linkHeader))
+			} else {
+				break
+			}
```
```suggestion
			linkHeader := resp.Header.Get("Link")
			nextEndpoint = fmt.Sprintf("%s%s", registryEndpoint, getLink(linkHeader))
```

kopiczko commented on pkg/quay/quay.go:
```diff
@@ -0,0 +1,166 @@
+			reposToSync = append(reposToSync, fmt.Sprintf("%s/%s", q.namespace, repo.Name))
+		}
+	}
+
```
Don't we have to do `resp.Header.Get("Link")` thing here?

corest replied:
For now, it doesn't give more than one page. So I can't even test it of there would be more repos to load

kopiczko commented on pkg/azurecr/azurecr.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+func New(c Config) (*AzureCR, error) {
+	if c.RegistryName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "RegistryName must not be empty")
+	}
+	if c.Credentials.User == "" {
+		return nil, microerror.Maskf(invalidConfigError, "User must not be empty")
+	}
+	if c.Credentials.Password == "" {
+		return nil, microerror.Maskf(invalidConfigError, "Password must not be empy")
+	}
```
```suggestion
	if c.RegistryName == "" {
		return nil, microerror.Maskf(invalidConfigError, "%T.RegistryName must not be empty", c)
	}
	if c.Credentials.User == "" {
		return nil, microerror.Maskf(invalidConfigError, "%T.User must not be empty", c)
	}
	if c.Credentials.Password == "" {
		return nil, microerror.Maskf(invalidConfigError, "%T.Password must not be empty", c)
	}
```

kopiczko commented on pkg/dockerhub/dockerhub.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func New(c Config) (*DockerHub, error) {
+	if c.Credentials.User == "" {
+		return nil, microerror.Maskf(invalidConfigError, "User must not be empty")
+	}
+	if c.Credentials.Password == "" {
+		return nil, microerror.Maskf(invalidConfigError, "Password must not be empy")
+	}
```
```suggestion
	if c.Credentials.User == "" {
		return nil, microerror.Maskf(invalidConfigError, "%T.User must not be empty", c)
	}
	if c.Credentials.Password == "" {
		return nil, microerror.Maskf(invalidConfigError, "%T.Password must not be empty", c)
	}
```
